ral-facilities/datagateway-api
# Search API Session/Client Handling

This PR will close #258 

## Description
This PR adds a solution for client handling for the search API. Compared with DataGateway API, this is a much more basic solution; because the search API will only ever use the anon user in ICAT, I think a single client object can be used for the API. This client is stored in a `SessionHandler` class which is accessed by the endpoints when they need it. 

The session for this client is managed by the `client_manager` decorator. This is a similar design to `requires_session_id` on DataGateway - i.e. a decorator to be used for endpoints. The client manager decorator looks at the remaining lifetime of the session and if the session has expired, the client will log back in as an anon user. There's no initial login/session ID being set as this use case (i.e. the first time `SessionHandler.client`) is used, a `ICATSessionError` will be caught due to the empty `sessionId` attribute in the client.

I've made use of `ICATClient` for the search API to reduce duplication - this class inherits Python ICAT's `Client` but extends it slightly for the cache object pool design. `client_use` now allows for client configuration for both search API and DataGateway API. 

In terms of testing, I've done some mocking of config (really to test that `client_use` does what I expect it to) so I've moved a couple of pytest fixtures out to `conftest.py`. I've also had to put the config into its own class in `config.py` because otherwise, it cannot be mocked. This is because when `config.py` is imported, the old way (i.e. no in a class) is set as soon as its imported which makes it impossible to mock.

The commits tell a 'nice story' (they're not all over the place and confusing like usual!) so I won't go through each change one-by-one, hopefully looking through each commit gives you a good understanding.

This is a new 'feature', so the new version of the search API will be 3.1.0.

## Testing Instructions
I've done some testing using Apache JMeter and it passes, with no drop in speed (all requests were <600ms on my machine) - 200 'users' are spun up and each send a request to `/datasets`. I'm happy to give the file to the reviewer if this helps.

It should be noted that the ICAT anon authenticator is required for all the tests to pass. This is because `SessionHandler.client` will login using that authenticator. I've made sure the CI has this authenticator by [making a change in ICAT Ansible](https://github.com/icatproject-contrib/icat-ansible/pull/71).

## Agile Board Tracking
Connect to #258
